### ISSUE TITLE
Request for position on Serial API

### ISSUE THREAD

**Issue by computersarecool** [mozilla/standards-positions#336]
## Request for Mozilla Position on an Emerging Web Specification

* Specification Title: Serial API
* Specification or proposal URL: https://wicg.github.io/serial/
* Caniuse.com URL (optional): NA
* Bugzilla URL (optional): https://bugzilla.mozilla.org/show_bug.cgi?id=webserial
* Mozillians who can provide input (optional): @marcoscaceres 

### Other information
According to https://github.com/WICG/admin/issues/108#issuecomment-626475273 
there is no input from Mozilla. I hope Mozilla does provide input as this spec would be a useful addition to moving the web forward.

**Comment by dbaron** [comment]
Also cc @martinthomson.

**Comment by marcoscaceres** [comment]
cc'ing @reillyeon, who might be able to provide input on [Chrome's status](https://www.chromestatus.com/feature/6577673212002304).

This has very similar security characteristic as Web MIDI, Web BlueTooth, etc. so probably the same concerns apply.

**Comment by reillyeon** [comment]
An MVP of this API is available as an [Origin Trial](https://developers.chrome.com/origintrials) in Chrome, and the feature can be enabled with a flag in other Chromium-based browsers. The specification is still a work in progress, with the prototype in Chromium allowing me to flesh out the details beyond the initial API sketch @marcoscaceres provided in the original specification. Time permitting I plan to finish updating the specification based on the prototype this quarter.

**Comment by tomayac** [comment]
There's a [codelab](https://codelabs.developers.google.com/codelabs/web-serial/index.html?index=..%2F..index#0) written by @petele available to illustrate the mechanics of the API.

**Comment by martinthomson** [comment]
As @marcoscaceres points out, this shares a lot of characteristics with WebUSB (and friends).

It appears that the only security mechanism protecting devices from sites is user consent.  And that is only at a "RECOMMENDED/SHOULD" level.

We don't believe that user consent is adequate protection for anything that provides this level of capability.  Serial access is a relic from an age where a physical connection conferred a great deal of trust.  For instance, many devices offer administrative control to anything that connects over this interface without any form of authentication; in my experience, this often extends to privileges that transcend even what a root user can do.

Even with the [additional safeguards](https://wicg.github.io/webusb/#webusb-platform-capability-descriptor) in the WebUSB spec, we still regard that as `harmful`.  The dynamics are slightly different with serial than USB, but they seem at least comparable.  Even if we could establish that the risks are manageable in the same way, the protections are also considerably weaker.

I'm going to recommend we mark this as `harmful`.

**Comment by reillyeon** [comment]
This is a somewhat pedantic point but since this has come up in other contexts I feel it is worth mentioning here. [RFC 2119](https://tools.ietf.org/html/rfc2119) defines "SHOULD" as,

> This word, or the adjective "RECOMMENDED", mean that there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

In the area of permissions I feel that "SHOULD" is an entirely reasonable requirement. For example, both Firefox and Chrome provide enterprise policy controls which allow an administrator to make policy decisions on behalf of their users. Chrome supports this for WebUSB permissions and would be out of compliance if the specification said that it "MUST" prompt the user.

**Comment by kgiori** [comment]
@dbaron -- I wish that I could understand this "harmful" position that Mozilla is taking for web serial, but I cannot. Am I wrong in presuming that such a position will likely keep Firefox from implementing it? If so, then K12 education in the area of "physical computing" (programming microcontrollers for data science collection, robotics, mechatronics, electronic music and art, STEM, and IoT) will need to run Chrome. I feel sad that educators will not have a choice of browser when it comes to teaching these important new technology subjects. :(

To point to a specific example, look at recent comments (May 25) to the webserial [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=926940) from jmaloney to better understand why web serial is so important. John created a recent [video](http://microblocks.fun/mbtest/media/Robolot%202020%20MicroBlocks%20Keynote.mp4) of cool 0.5 MicroBlocks features that shows how amazing this project is. The video shows how some microcontrollers can interact with the web as http clients (~14 min) or servers (~15 min), and how you can stream images and sound to them (~17 min). MicroBlocks also supports a "Web Thing" library that enables microcontroller boards to be directly managed and monitored (just like commercial IoT devices) by the [WebThings Gateway](iot.mozilla.org/gateway) by Mozilla (where the gateway's web thing API is close to the W3C Web of Things standard). These two projects work really well together, and are examples of amazing new tools for education that rely on the open web, and web standards.

Supposing users were allowed to enable web serial as an extension in Firefox, is there anything that could be done or run on the microcontroller side of things to help reduce whatever security risks there are? Can anyone provide an example of a product or service that might pose a risk to me if I was able to use a web serial extension in Firefox? As you can see, I'm looking for ways to overcome or work around the security risks to Firefox users so that I do not have to give up on using it... :(

**Comment by bromagosa** [comment]
I genuinely don't understand the risks either. What is it that is so harmful about serial access vs, say, the privacy risks involved in accessing cameras and microphones from the web, also secured "only" behind user permission?

I may be overlooking something, but WebSerial involves devices that have a serial interface, which I believe rules out pretty much every single end-user commercial USB device and puts this standard into a totally different category than WebUSB.

**Comment by fabricedesre** [comment]
The main issue is that it's impossible to get informed consent from the user with just a permission prompt for an api like WebSerial. It's pretty clear what the consequences can be when you allow a page to use your camera or microphone, much less so with an api that exposes a low level protocol.

Ultimately, this is an issue with establishing trust: how do you trust the page to not serve code that will damage your equipment? It's not enough to trust the developer, because its infrastructure can be compromised to serve harmful code. Web of trust doesn't help much either since you can target different payloads to different people. SRI suffers from a "trust of the root" bootstrapping issue.

Native platforms rely on code signing by their store: users trust the store editor to properly vet the app.
This is also what FirefoxOS/KaiOS do with signed packaged to grant powerful permissions. Here the drawback compared to regular web pages is that you can't easily link into packages (they need to be installed first) and you may have to wait for updates to be approved.

That puts the web in a tough position. Maybe a solution would be to layer "store like" signing on top of web packages to preserve urls, or only allowing web-extensions to use these apis and verify these extensions are safe?

**Comment by HenrikJoreteg** [comment]
Just had a lengthy discussion with @marcoscaceres about all this on twitter. Twitter doesnt' make it easy to share it all, but it starts here: https://twitter.com/marcosc/status/1296677919813582848

Going to pull in relevant pieces here:

**Marcos:**
If you tell me, "I can't build my business because my users can't do X", then I'm all ears. But, cases like "be cool if I had access to X because I want to fool around with X... not so much".

**Me:**
This isn't a "would be cool" scenario.

Quick background: I built http://AnesthesiaCharting.com with a partner. We help doctors/dentists/oral surgeons/CRNAs produce "sedation records" which is a specialized kind of medical chart that you have to produce when you anesthetize someone.

**Marcos:**
Go on... what can't you do?

**Me:**
Biggest thing customers wanted after we built/launched our initial version was to pull patient vitals (heart rate, blood pressure, etc) directly off of vitals monitors.

These vitals monitors are not typically very modern, standardized, or easily upgraded. 

Many of these that are out in the wild only have simple serial interfaces.

Or require local TCP connections.

**Marcos:**
Ok, you are hitting on the problem. These devices were never designed to interface with an extremely hostile software environment. 

So you need some kind of socket connection?

**Me:**
At minimum I would love to be able to initiate a TCP connection to a known port. Ideally I would also be able to open a server and listen for incoming connections. I realize the latter is more complex.

My customers don't want to install anything. I don't want to maintain and ship updates to a distributed piece of software for multiple platforms.

And if I could just get this to work on the web... all these problems go away.

This is a huge deal to me. 

This project is not some giant, corporate thing. My business partner and I realized we could possibly save a few lives from accidental deaths via mistakes that occur in anesthesia procedures.

Of course I want to make money with this, but only because I want to be able to work on it and similar problems.

Having a distribution platform like the web that has these types of APIs opens up whole new worlds.

**Marcos:**
Let's just spitball here... you could connect the equipment to a server, then have the sever authenticate users, then proxy commands over web sockets... for instance.

**Me:**
I did exactly that, actually. I wrote a node gateway server that would sit on the local network and broadcast data back to my API. 

Think about what this does to my business model.

Instead of distributing software on the web... now I'm distributing hardware?!

This also sucks. Distributing/maintaining software updates to a distributed piece of hardware living on these local networks... Is a completely different game with a completely different set of constraints and costs.

If I want to be able to make a business out of this and keep costs low for customers thereby lowering costs of medical care, I need a way to make these connections.

What's the real risk of allowing a user to allow permission for me to connect to a specific IP/port they approve?

**Marcos:**
It's the user that approves, not the machine... the machine doesn't assume you are hostile. You can then manipulate the machine, set it on fire, or do whatever you want if it's not expecting a hostile actor.

**Me**
There's nothing new here. If I'm running software on your local network I can do whatever the hell I want. 

The local gateway/server idea is not more secure.

This is a privilege every native app enjoys, no?

At least here we're talking about explicit permission.

The alternative here is that somebody does the exact same thing I'm doing with an iOS app or a really crappy Windows app.

Again, this is not more secure.

Just go ask oral surgeons how they do it right now 😅 

**Marcos**
That's not a net loss... the web has always lived alongside native apps and iOS/MacOS can be pretty secure (it hosts web browsers after all). 

I'm not saying I'm giving up here. The responsibility of privacy/security needs to be shared amongst all parties.

**Me**
What it does do, IMO... is limit who gets a shot it fixing these types of problems.

There's a whole other level of cost/complexity involved if you have to distribute/update/maintain software for a plurality of platforms. The magic of the web is that it removes so many barriers.

Centralizing API access through the web is an incredible tool for leveling the playing field for humanity as a whole (going back to Mozilla manifesto).

Instead of needing 3 full teams of devs for 3 years I can ship an app myself that potentially lowers medical cost for many.

*That* is a big deal to me.

This is about *access*....and wow as you mentioned... think about education.

Holy shit! Can you imagine how much amazing stuff could be produced if we lowered the barriers to entry like this? Would be incredible.

**Marcos:**
*I personally* think it's a game changer. But I don't represent Mozilla or their position on the matter (link below for those following along).  

And I also understand what Mozilla's concern are. They are valid.  

**Me:**
They are valid in isolation.

I believe if you look at the *whole* threat model... that the alternatives are *worse*. 

In terms of holistic impact... preventing evolution of these APIs is just leading to crappy workarounds.

I really appreciate you hearing me out.

As I said in another conversation this week... disagreements just mean that people care.

What I'm trying to provide is a concrete example of a real-world use case that I believe is significantly improved by adding these APIs to the web ❤️ 
https://mozilla.github.io/standards-positions/#webserial

**Comment by HenrikJoreteg** [comment]
for additional context, the final outcome of these not being available is me building/maintaining an Electron app. I first built an iOS app that was rejected for not using IAP. 

Distributing cross platform software *sucks*, by which I mean, it's prohibitively complex, expensive, and time consuming. (EV code signing anyone? https://twitter.com/HenrikJoreteg/status/1295563511863549953).

Anything that is possible through the web, **dramatically** levels the playing field and lowers the barriers to entry.. which i feel would ultimate help Mozilla live up to its manifesto. 

That is why I always have and will continue to [bet on the web](https://joreteg.com/blog/betting-on-the-web). I also wrote a whole [post about adding shiny new capabilities to the web](https://joreteg.com/blog/project-fugu-a-new-hope) and why I think its so important ❤️

**Comment by SuperUserNameMan** [comment]
> @HenrikJoreteg the threat is that code can be changed server side without the user noticing because you can't bootstrap a trust chain. From there, the scenarios are only limited to attackers imagination.

Very funny. If that's the way to think, then you should display a warning with an "uninstall Firefox to be safer and disconnect your device from the internet" button each time a user launch Firefox.

**Comment by andreicristianpetcu** [comment]
@SuperUserNameMan do you think a webextension api is more secure, less secure or equally secure to a web api?

**Comment by SuperUserNameMan** [comment]
@andreicristianpetcu :
> @SuperUserNameMan do you think a webextension api is more secure, less secure or equally secure to a web api?

I don't care.

I just don't want Firefox to become the Internet Explorer of the future ( ie : the navigator that slows down the internet evolution and make innovation a PITA just because the devteam don't want to follow the trend imposed by others and because of an irrational fear of the "unlimited imagination of some imaginary hackers" who already plague the web using simple emails with an html attachment).

Also, unlike Internet Explorer (and the new Microsoft Edge) that was preinstalled on almost every computers, installing Firefox is purely optional or is almost a political choice. As the Serial API is slowly becoming a must-have in some growing DIY electronics communities and in more and more schools, the refractory and aging Mozilla Firefox will slowly lose market share by becoming an very optional relic of its own glorious past. Not mentioning that Google Chrome is the perfect extension to Android devices that almost everybody on Earth have.

A web extension would certainly make Firefox "safer" for most of its user base since installing the extension would require a relatively complicated and conscientious action from the user, but will it prevent the unlimited imagination of the hackers and scammers from hacking and scamming ? No, it will just make the life of some Firefox users a little bit more complicated, and little bit more inclined to use Google Chrome (which can be installed on Linux BTW) .

**Comment by mlyle** [comment]
I'm just here to note that there's a whole lot of educational use cases.  It's hard to get students to install extensions, Electron apps, etc.

I teach a middle school class that I've written a web-based IDE for-- it talks to embedded micropython development boards with webserial.  I like Firefox but effectively I am put in the position of pulling people over to Chrome, instead.

**Comment by jamesturner246** [comment]
What's wrong with adding an option to `about:config`? That way there's little chance of a lay user accidentally activating it, while the users that wish to are implicitly competent and aware of the risks? There's even a giant 'Proceed with Caution' warning for free.

**Comment by mmiscool** [comment]
It seems like a mistake to lump this in to a category of low level hardware access. Its not especially low level to access a serial port and the positive interaction from the user to select the specific serial device seems like more than enough. Why force some one to possibly install a less secure thing like an extension or other program that actually needs to run outside the browser sandbox to make the user interact with a serial device. I believe that the decision to label this harmful is a mistake and that the excuses as to why it is a security concern are week at best and disingenuous at most. Forcing a user to install some thing outside of the browser sandbox is dangerous and the positive implications of this kind of API are clearly evident for the maker community, schools and hardware developers. Serial is an ancient technology and making it easy to work with for the development platform of the future seems like a short sighted decision.

**Comment by kgiori** [comment]
> I'm just here to note that there's a whole lot of educational use cases. It's hard to get students to install extensions, Electron apps, etc.
> 
> I teach a middle school class that I've written a web-based IDE for-- it talks to embedded micropython development boards with webserial. I like Firefox but effectively I am put in the position of pulling people over to Chrome, instead.

The ability to run [MicroBlocks](http://microblocks.fun/) in the browser -- http://microblocks.fun/mbrun -- is similarly driving learners of physical computing to Chrome, due to its web serial support. Sad to see Firefox not supporting education.

**Comment by fabricedesre** [comment]
The argument of "Forcing a user to install some thing outside of the browser sandbox is dangerous" just ignores that the same API in the browser would not be sandboxed either: a device chooser doesn't protect you against malicious code being served.

Exposing this kind of APIs to extensions only - since they are signed, can be revoked etc. looks like a better solution to me (disclaimer: I have no weight on Mozilla's decisions).

**Comment by sbrl** [comment]
I appreciate it's is potentially dangerous, but at the same time there are a lot of excellent applications here. The question I think here is how to communicate the risk effectively to users? And also how to limit the risk of malicious code being run?

What about requiring [Subresource integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) and/or a content security policy or something?

**Comment by fabricedesre** [comment]
SRI still needs something to bootstrap the chain of trust, to be sure the page linking to the subresource has not been tampered with.

**Comment by crozone** [comment]
Mozilla's stance on this is still incredibly frustrating. This is not USB access, it is not low level device access. It is a feature that allows a user to consent to a web page sending and receiving text out of a single, specific, user specified hardware port. The user chooses exactly what serial port to use, not the website. It's really a high level API in the scheme of things. In terms of risk characteristics, I still fail to see how serial access is more dangerous than microphone access, camera access, or even allowing a user to download a file.

This feature is becoming more and more important in educational settings with MicroPython microcontrollers communicating with Web IDEs. Currently users must use Edge and Chrome to accomplish this. And guess what, users *will* just use Chrome and Edge for this purpose, and Firefox will not be used in those settings. Users like me will have to run both.

So on the topic of security... if a malicious website asks a user for serial port access, and then the user happens to have their heartrate monitor plugged in, and the user then selects the heartrate monitor, sure, that could be bad. But newsflash: If a malicious website asks you to download a random .exe, and the user runs that, the consequences are going to be far worse.

At some point, you're going to have to respect that users need to be able to make their own decisions regarding risk, Having a consent dialogue, just like **every other browser that implements this**, has proven to be perfectly adequate. In addition, the overlap of Firefox users and people who need this serial feature, are *more* than capable of assessing that risk themselves anyway.

Just my 2c.

**Comment by Phoenix1747** [comment]
> But newsflash: If a malicious website asks you to download a random .exe, and the user runs that, the consequences are going to be far worse.

While we're at it, we should disable the download feature so that people don't keep downloading virus.exe to their heart rate monitor.

**Comment by kgiori** [comment]
Here's an EdTech example for programming microcontrollers -- my favorite educational programming and learning tool is MicroBlocks:
[https://microblocks.fun](https://microblocks.fun)
Of course it runs directly in Chrome or Edge but not in Firefox. 
Can Firefox at least create or enable the user to create an "allow" list of educational websites like MicroBlocks to be allowed to run WebSerial?

**Comment by HenrikJoreteg** [comment]
In reality, if Firefox doesn't support it, nothing changes for us.

Those of us who need these features will simply continue to tell of it customers to download Chrome or Edge if they want to use our software.

That's the ultimate impact of not supporting these features. And no, that is not a hypothetical, that is literally what [we tell our customers](https://xchart.com/automatic-monitoring) and so far we have gotten precisely *zero* complaints about it.

In fact, that linked page is a good example of the ultimate impact of not supporting these APIs.

I'm not sure Mozilla ultimately protects anyone by holding this stance.

**Comment by mmiscool** [comment]
My position on this has changed. The right answer is just to detect if the user is using firefox and prompt the user to install a browser not controlled by paranoia.

**Comment by thorrak** [comment]
> My position on this has changed. The right answer is just to detect if the user is using firefox and prompt the user to install a browser not controlled by paranoia.

This is what I currently do with my project that uses Web Serial.

**Comment by NODeeJay** [comment]
> I'm not sure Mozilla ultimately protects anyone by holding this stance.

They are protecting Chrome, Edge, Opera and even the Samsung Internet browser. It's like parents which don't want to take the time to educate children but prefer to forbid something, the easy way out but not on the long run.

Maybe Mozilla, it's time to step down from the arrogant horse of authoritarian overpatronizing your users and start listening to them, as long as you have users left because you force them to use other browsers.

@fabricedesre sure, the tread is that every page can do what it wants. Driving is dangerous, that's why we have rules. Limiting a certain service to access only certain sites is a solution, NoScript effectively does it that way. Maybe the certificate can also be checked for that connection, oh wait, we do that already on any site. 
I am actually more concerned about your partnership with Meta, the privacy friendly company from Mark Z. But that's none of my business. Oh wait, as a user it is. And recently a new FF offers **Yandex.RU** as a search engine, this is most probably from my current point of view not just "problematic", but hey, WebSerial is dangerous.

@tantek the fact that users don't stop writing in this thread should show you enough about their loyalty towards this browser. Obviously it is the wrong place as usual and surely the questions is asked wrongly and OP surely has some typos in his post and most probably it's the wrong color...
In fact they want to help you to keep a user base at all, they care about you. They are interested in a bright and secure web future with a good browser. It's not that WebSerial is a nice cherry on a Sunday, it's a need. 
As @gniezen pointed already out: if you don't fulfill it, users **have** to change, **YOU force** them to move on to a less secure browser. So in the end this position does not help, it harms your users because the options we have are knowingly way worse (Chrome, Edge, Samsung Internet, Opera). But at least you still get around $ 400 mio from google (https://www.zdnet.com/article/sources-mozilla-extends-its-google-search-deal/) yearly, I know, my 100 bucks donation last year are not in comparison.

Firefox went down from 38.1% in 2010 to under 3% today (https://www.w3counter.com/trends), please stop trying that hardly to leave all up for Chrome in the end, please, we need you Mozilla, with WebSerial and WebUSB!

**Comment by fabricedesre** [comment]
> It's like parents which don't want to take the time to educate children but prefer to forbid something, the easy way out but not on the long run.

A more apt metaphor would be that chrome is like parents who let children do whatever they want, at the price of some of them getting hurt, while Mozilla prevents them all from being hurt until a solution is found to protect them all. The issue is that Mozilla is not trying much to find that solution imho.

> the tread is that every page can do what it wants. Driving is dangerous, that's why we have rules. Limiting a certain service to access only certain sites is a solution, NoScript effectively does it that way. Maybe the certificate can also be checked for that connection, oh wait, we do that already on any site.

No, the threat is that an attacker could change the code served by goodsite.com into something problematic.

**Comment by NODeeJay** [comment]
> > It's like parents which don't want to take the time to educate children but prefer to forbid something, the easy way out but not on the long run.
> 
> A more apt metaphor would be that chrome is like parents who let children do whatever they want, at the price of some of them getting hurt, while Mozilla prevents them all from being hurt until a solution is found to protect them all. The issue is that Mozilla is not trying much to find that solution imho.

I agree, or not willing to.

> > the tread is that every page can do what it wants. Driving is dangerous, that's why we have rules. Limiting a certain service to access only certain sites is a solution, NoScript effectively does it that way. Maybe the certificate can also be checked for that connection, oh wait, we do that already on any site.
> 
> No, the threat is that an attacker could change the code served by goodsite.com into something problematic.

Correct. As he/she could change the code when I download the bin/hex/... file to upload to an ESP. This threat is usually prevented by adding a hash verification, which still does not prevent the attacker from creating a new hash and posting it on the site, which still does not prevent the attacker from doing:

- DNS spoofing
- hacking the mailserver and impersonating as a goodcompany
- calling you in a social engineering attack as goodcompany
- posting fake information on youtube
- and so on.

Life itself is unfortunately the biggest threat to our lives, still we take risks. What I want to say is that there are so many risks out there that ultimately only our own death can protect us from them, still we live and take the risk everyday. When Mozilla forces its user to change to a less secure browser, than Mozilla is the the harmful threat, repectively those who took the decision there - we are also responsible for the things we do not do.

But there is an easy solution:
- about:config
- yes I understand the risks.
- Activate WebSerial, WebUSB, Web....
- only activate for SSL sites
- ignore invalid certificates
- only activate in local network
- only activate for whitelist
- activate for all but blacklist

and I am ready to write a proper documentation that scares the shit out of the user who wants to flash his ESP at home. Or the poor factory worker that gets instructions by the manufacturer to quickly FW upgrade his CNC machine that he continue his work. Or the diabetes person who has an issue with his pump and the support personnel is trying to debug the issue. Or the children in school which make the LED on the board blink through MicroBlocks.
I really don't want to image what would happen if an attacker would let the LED blink red instead of green...

This is from my POV a new level of ridiculousness and arrogance.

Anyway, plenty of proper solutions were named, not just in this forum, also on hackster, reddit and other corners of the internet. Mozilla needs money to run their servers and provide the service. Mozilla gets mainly paid by Google/Alphabet for installing the search engine as default. This is surely calculated based on the amount of downloads/installs/etc. We still 2.6% loyal users warned Mozilla enough, Darwinism will resolve the issue.

Mozilla wants to be like Netflix and Kodak, though times are ahead but then it will be better - hopefully.

**Comment by NODeeJay** [comment]
> > My position on this has changed. The right answer is just to detect if the user is using firefox and prompt the user to install a browser not controlled by paranoia.
> 
> This is what I currently do with my project that uses Web Serial.

Do you mind sharing some code. JS or whatever? This could be used as a real solution now.
What I have in mind, when users go to that site and they have the right browser installed, it starts the site in the right browser that has WebSerial enabled. I saw this feature with Edge when I open pages in IE that are not supported but I did not dig into it yet.

**Comment by NODeeJay** [comment]
Additionally, based on https://github.com/mozilla/standards-positions/blob/main/CONTRIBUTING.md I am hereby asking Mozilla / @martinthomson to update the position on this topic.

Why:
Many solutions have been provided in the meantime to fulfill this request in a secure way, and also many ways of securing the access itself. Hiding these settings for the experienced users is a better way of implementing it, than no implementation at all, for example:
- about:config
- yes I understand the risks.
- Activate WebSerial, WebUSB, Web....
- only activate for SSL sites
  - ignore invalid certificates
- only activate in local networks 
- only activate for whitelist IP/hostname
- activate for all but blacklist IP/hostname
- or generally the same list of exceptions like for "unsafe" pages

Furthermore, I am pointing out your core values:
Our mission is to ensure the Internet is a global public resource, **_open_** and accessible to all. An Internet that **_truly puts people first_**, where **_individuals can shape their own experience_** and are **_empowered, safe and independent._**
Most important here is that you **_empower us in a safe and independent way_**!

I therefore kindly urge you to revise your stance on this topic.

Thank you.

**Comment by thorrak** [comment]
> > > My position on this has changed. The right answer is just to detect if the user is using firefox and prompt the user to install a browser not controlled by paranoia.
> > 
> > 
> > This is what I currently do with my project that uses Web Serial.
> 
> Do you mind sharing some code. JS or whatever? This could be used as a real solution now. What I have in mind, when users go to that site and they have the right browser installed, it starts the site in the right browser that has WebSerial enabled. I saw this feature with Edge when I open pages in IE that are not supported but I did not dig into it yet.

Here's the code that I'm using in my project: https://github.com/thorrak/brewflasher_web/blob/e6b152c0a676845edfafc1d8248ec2078780fdbf/src/FirmwareDirect.vue#L69

**Comment by ocdtrekkie** [comment]
As a note, really glad to see Mozilla holding their ground here. I know it's rough to make the right call when so many selfish and lazy developers are demanding you put billions of users at risk to serve their thousands of niche tech users.

Positions like this continue to be why Firefox is the wise choice for enterprise and general consumers alike from a safety and security standpoint.

Web Serial remains a proprietary API with a single implementation written by a company with an active interest in harming consumer privacy and security. It isn't a standard, and we should stop acting like it ever will be.

**Comment by mlyle** [comment]
@ocdtrekkie-- I find your comment laughable.

> As a note, really glad to see Mozilla holding their ground here. I know it's rough to make the right call when so many selfish and lazy developers are demanding you put billions of users at risk to serve their thousands of niche tech users.

I am a middle school teacher.  I just want to stop telling students in my class to install Chrome.  I have an IDE that lets them do microcontroller development from the browser.

> Positions like this continue to be why Firefox is the wise choice for enterprise and general consumers alike from a safety and security standpoint.

Asking the user whether to do it is pretty damn safe.

> with a single implementation 

Chrome, Edge, and Opera-- all Chromium derived, sure, but all support it.

> written by a company with an active interest in harming consumer privacy and security.

Yup.  And it doesn't harm consumer privacy or security.  On the other hand, the lack of it in Firefox makes me move students to the platform that does harm consumer privacy and security.

80%+ of my students already have Chrome.  Moving the 20% of them that don't to Chrome is more tenable than installing something dodgy to do development on 100% of machines, even if I find it icky.

**Comment by ocdtrekkie** [comment]
> I find your comment laughable.

Great way to start your reply.

> I just want to stop telling students in my class to install Chrome.

Then stop. That's entirely your decision to harm your students.

> I have an IDE that lets them do microcontroller development from the browser.

I recommend finding an IDE that doesn't require the modern equivalent of ActiveX to work. Proprietary Chrome features come and go every day.

> Asking the user whether to do it is pretty damn safe.

This is kinda where you really went off the rails though: This tells me you don't know much about user interaction with technology, which like... you should if you're teaching microcontroller development in middle school. Please pay attention here, and let me help: **Users will click literally anything that lets them proceed, and likely, won't even remember doing it.** This is true in schools, in enterprise businesses, and in homes. It's *universally* the case that asking the user if they want to do it is *not at all safe*.

> all Chromium derived

As I said, a single implementation written by Google. Edge and Opera didn't reimplement or likely even meaningfully vet this code, they're just in the business of running Chrome forks.

> And it doesn't harm consumer privacy or security.

False, and both Brave and Mozilla browser security teams disagree with you. (And understand user safety more, see above.)

> Moving the 20% of them that don't to Chrome is more tenable than installing something dodgy

I'm not sure what you'd refer to that could possibly be more dodgy than Chrome. To be clear: Chrome is the *only* browser refusing to block third party cookies going on a few years now, because they won't do it without implementing a different way to invade user privacy first. Chrome is, by any reasonable definition, adware, and should probably be blocked by your security software. It is by mine.

**Comment by mlyle** [comment]
> Great way to start your reply.

Your opening salvo was to refer to us as "selfish and lazy developers".  If you didn't want a flamewar, maybe you should have picked a gambit that allowed room for reasonable conversation.

> I recommend finding an IDE that doesn't require the modern equivalent of ActiveX to work. Proprietary Chrome features come and go every day.

I wrote it.  It's a much less user-hostile thing than anything else I found for students to install.

> Users will click literally anything that lets them proceed, and likely, won't even remember doing it.

Speculatively popping up a "pick a serial device" dialog on lots of peoples' machines-- in hopes that they find they correct device you want to exploit, select it, and click "permit" doesn't seem like the most likely attack scenario to work or the low hanging fruit.

> Chrome is, by any reasonable definition, adware, and should probably be blocked by your security software. It is by mine.

Ah, we've got Quixote here wanting to prohibit the thing with 70% market share from end-user controlled devices.  That's a practical plan.

**Comment by ocdtrekkie** [comment]
@mlyle Ah. So you wrote the bad software, and want to force billions of users to have a more vulnerable platform to cater to it. That explains it much clearer, I appreciate that.

The core problem here, is that yes, users will absolutely select an enumerated device and let a random website connect to it. How do I know? I see stuff that dumb *every single day*. Browsers need to support people from ages 6-106 as LEGO like to put it. Your students can install a piece of local software, and I bet you can even write that too.

Also, one of the recent issues has been attacks on industrial control systems. In one example, a site offered a tool explicitly to help ICS operators, which also took malicious actions. In that case, a user would absolutely authorize it, and they'd permit malicious activity. Of course, here's the key point: It is drastically easier to monitor, secure, and block separate software and processes. Hiding this inside chrome.exe makes it very hard to secure... which is actually Google's primary motivation for injecting so many features directly into the browser. Following Google's lead here is to intentionally make other browsers as compromised as Chrome already is.

**Comment by mlyle** [comment]
> Ah. So you wrote the bad software

I did my best as an engineer and teacher to find something that was low impact that would allow my students to do their work.  If you want to sneer at me about that, then you are just not being very nice.

I still don't know of a better option.  I wasn't super excited about writing an IDE. and I wasn't super excited about requiring Chrome.  In the real world, we don't get exactly what we want.

>  In one example, a site offered a tool explicitly to help ICS operators, which also took malicious actions

So, the users downloaded and ran a tool and explicitly authorized it to talk to controls.  Perhaps Firefox should not allow downloading programs to preclude this type of attack.

> Of course, here's the key point: It is drastically easier to monitor, secure, and block separate software and processes.

It is somewhat easier.  Of course, if the real intent is to prevent malware from talking to the industrial control system, perhaps the better angle is to use an explicit allow list for what can instead of giving everything on the computer permission and hoping users don't manage to run the wrong thing.

>  Following Google's lead here is to intentionally make other browsers as compromised as Chrome already is.

I've always thought that it's important in a conversation to presume good faith.  Any argument that the engineers on Chrome are actively trying to make the web a less secure place to somehow make Chrome look better leaves no room for discussion, and doesn't seem particularly likely-- and it kinda fails on Occam's Razor.

**Comment by ocdtrekkie** [comment]
> I still don't know of a better option.

I would probably either use one of the numerous cross-platform UI toolkits that lets you use HTML for the frontend, or stand up a quick web server on localhost so that it works in any browser. I think there's a lot of solutions here that don't involve proprietary browser features.

> It is somewhat easier.

It is *significantly* easier to monitor external processes than what the dozen some-odd Chrome processes are all trying to do on a PC. In my environment I have a dozen ways to monitor, block, and restrict processes. Web traffic is far more difficult to restrict as well, especially if you don't disable all of Chrome's security-defeating protocols. Obviously, in the office we generally use policies to cripple nearly all of these antifeatures, but it's a constant game of whack-a-mole because Google adds a new one every time someone wants a promotion and a cool demo for their Google I/O talk.

> I've always thought that it's important in a conversation to presume good faith.

It is. However, once someone's shot you in the head, you don't have to assume they're in good faith anymore. We are several years past the point where Google had numerous opportunities to make ethical decisions that help make the web more secure, and they have chosen ad revenue *every* **single** *time*. It's nearly impossible to have a browser standards discussion without the understand that Google is absolutely acting in bad faith, and that they will do anything they can to harm the open web.

**Comment by mlyle** [comment]
> I would probably either use one of the numerous cross-platform UI toolkits that lets you use HTML for the frontend, or stand up a quick web server on localhost so that it works in any browser. I think there's a lot of solutions here that don't involve proprietary browser features.

One to one, and student owned devices.  You're describing me running a service on 30+ student owned computers, including a few Chromebooks.  I won't get parent permission for all of them.  It will not work on all of them and I'll get stuck for hours doing support.

In the end, Chrome is the closest thing to universal in the student environment; everyone has it, even if a few prefer Safari or Firefox.  And every thing that Firefox does like this in the name of "sek00rity" makes it harder for those preferring other browsers to keep using them.

> and they have chosen ad revenue every single time.

Serial doesn't benefit their path to ad revenue-- well, only indirectly.  When Firefox forces many of us to use Chrome to get the feature, they get our eyeballs.

**Comment by mlyle** [comment]
@ocdtrekkie In the end, bruh: I'm not a dummy, selfish, or lazy.  I know a whole lot about security and development, and implying anyone who disagrees doesn't is an asinine way to make an argument.  I'm trying to bring really deep undergraduate level curriculum to the middle school environment, and **very reluctantly** writing a whole lot of tooling and doing a whole lot of desktop support to make it happen.  There are very few perfectly tidy solutions that work.

**Comment by ocdtrekkie** [comment]
Well, neither userspace apps with cross platform UIs nor serving web content on a high number port requires admin access...

> including a few Chromebooks

And there we go, now we've come to the truth. The only solution that works on proprietary Google products is a proprietary Google API. And everyone need be compromised to support Google's netbook platform.

Again: Users will click literally anything to proceed. Chrome is a veritable soup of tracking and malware vectors, a feature like this is dangerous and poorly warranted for the relatively niche crowd of microcontroller developers. I don't think the cost of implementing this antifeature is worth the benefit of a handful of Chromebook users being able to play with a micro:bit. I hope Mozilla holds firm here.

**Comment by mlyle** [comment]
> And there we go, now we've come to the truth. 

Not very many Chromebooks-- even without them, we would still get pointed this way, though.

**Comment by SuperUserNameMan** [comment]
Oh ok cool ...

So basically, so maybe Firefox should be advertised this way instead :
----
![Screenshot 2022-07-31 at 10-22-28 Download the fastest Firefox ever](https://user-images.githubusercontent.com/9801802/182017318-7279b187-3cc1-4148-be09-1062199fd406.png)

----

Because else, I wonder why dumb lazy people would bother installing Firefox (or even Google-Chrome) by themselves when they already have a default web-browser available on their Windows, MacOS or Android devices that perfectly does every dumb thing they already need to do on the web.

Willing and being able to install Firefox seems a pretty over-complicated dumb thing to do for a dumb lazy web-surfer ... why in the hell would they want to do that ? Even Microsoft Widows tries its best to prevent this behavior ...

Wouldn't it require a certain level of awareness and intelligence ?

And what does it tell about all these Linux users who ends-up with Firefox preinstalled on their favorite niche distro ?
I guess they must really be this kind of dumb lazy users who would click anything that pop up on the screen just to proceed to the next page ...

**Comment by ocdtrekkie** [comment]
@SuperUserNameMan You sure put a lot of effort into flaunting your personal superiority over the security teams of *every single web browser developer that isn't an ad company*.

**Comment by mlyle** [comment]
@ocdtrekkie You put a lot of effort into personal attacks and selective mischaracterizations to troll on GitHub.

**Comment by NODeeJay** [comment]
Dear @ocdtrekkie, it is in my POV very sad how you jeopardized this still technical conversation. You ignore all good rules of a rational debate (https://en.wikipedia.org/wiki/Wikipedia:Rational_debate) including but not limited to:

_ 1. "[...] why Firefox is the wise choice for enterprise and general consumers [...]" implies that all others are dumb & "[...] So you wrote the bad software [...]" just to name a few. 
_ 2. "[...] This is kinda where you really went off the rails though: This tells me you don't know much about user interaction with technology, which like... you should if you're teaching microcontroller development in middle school. Please pay attention here, and let me help: [...]"
_ 6. "[...] I recommend finding an IDE that doesn't require the modern equivalent of ActiveX to work. [...]"
_ 5. 7. & 9. "[...] How do I know? I see stuff that dumb every single day. [...]"
This is the one point where you are right. Fully right. How often I was driving km in the evening to press a button, that 4 other people assured me to have pressed before. After 30 years in IT I am more than capable of writing a joke book about it. But, do we forbid the user now to browse the internet or use computers? Your job would be pretty boring if we would do so. We put safety belts and airbags and auto break systems into cars to help the users continue in a safer way, this is the solution. Otherwise the war on security is lost the same way as the war on drugs. Btw. drugs won.
_ 10. "[...] you don't have to assume they're in good faith anymore [...]"
Just the tip of the iceberg I saw.

You failed in doing your background check, even with LEGO, it's from 6 - 99, not 106 :-)

You also failed to read other suggestions and comments before, e.g. my sum up https://github.com/mozilla/standards-positions/issues/336#issuecomment-1173177328 & https://github.com/mozilla/standards-positions/issues/336#issuecomment-1173143179 which would, backed by sources, clearly shock you as Mozilla sold you to google already, the bad company you are pointing out here. Not saying that I am a friend of google, but they are there and they at least they do some job, not like Mozilla by just saying: Dangerous, we don't do.
Instead you decided to let your bad mood or bad day drive your need for totally unneeded comments in a technical forum.

You also failed in providing any source for your opinions, in no comment you come up with any verification, source or further reading to proof your point.

In https://github.com/mozilla/standards-positions/issues/336#issuecomment-1200282069 you properly ignore the requirements the poster pointed out. Sure, micro controller development can be still done with an application on the computer, or even by just copying the code to the serial port - old school but works. The thing is, when we want a next generation that is not forced to purchase Nest, Tuya or closed source MCUs, then we should teach them how to DIY, thank you @mlyle for your service on our civilization.
You are right, the good people at brave which sold our data (you will find the link, use the bad company to search for it) and give us points for watching ads....

While @mlyle did great in trying to keep the conversation on a discussion and technical level, you properly failed.

I don't share @SuperUserNameMan style, but I certainly share his frustration and arguments and I am also one of these dumb lazy users :-)

@ocdtrekkie I would appreciate of we stop these defamations now and get back to a technical conversation. As you seem to be a very good programmer, utterly experienced in all corners of life and explained that you understand the dumbness of people thourougly, what would a proper solution to implement it in a way, that even the dumbest user is safe. Would be a whitelist an option or maybe limiting the feature to the internal network? What is in your POV a smart way to protect the users but still leaving the feature available?

**Comment by ocdtrekkie** [comment]
@NODeeJay Hardly, @mlyle came at the issue dismissively, with a complete lack of concern for security, and with the absolute bold arrogance to assume he knows better than Mozilla and Apple's combined security knowledge... because, well, it makes *his job easier*.

I understand you also disagree with me, but that's not an excuse to pretend mlyle had technical reasons and I didn't. I explained practical real world attacks, and mlyle responded by simply saying he thinks asking users permission is adequate safety, even though it's... very plainly not.

**Comment by SuperUserNameMan** [comment]
Maybe that would be easier to understand and accept if we were given actual examples of security issues related to the Serial port API.

- Does it happen often in Google Chrome ?
- What kind of actual devices are connected to the serial ports of most Firefox users ?
- Who still uses serial ports nowadays and for what purpose ?
- Who's actually at risk ?
- What kind sensitive devices would be connect to serial ports ?
- And if a highly sensitive device is connected to a serial port, who would be inclined to take the risk to use the same computer to browse insecure parts of web ?

**Comment by mlyle** [comment]
> @NODeeJay Hardly, @mlyle came at the issue dismissively, with a complete lack of concern for security, and with the absolute bold arrogance to assume he knows better than Mozilla and Apple's combined security knowledge... because, well, it makes his job easier.

@ocdtrekkie Obviously different vendors agree, and you agree more with Firefox's current choice.  It's not a great way to proceed to denigrate the entire other side.  You fired the opening salvo of mockery with your post in the thread and it's something that I've tried to avoid in talking to you-- showing you respect even as you refuse to do so for the people who disagree with you.

I actually spent the entire beginning of my career in computer security and had some notable accomplishments (analysis of malware, development of new types of attacks, new security technologies, patents, building 8 figure business groups and selling a startup for 9 figures; being well-regarded in security circles, etc.).  I don't like argument from authority, but you should keep in mind that when you're talking to people that they *may actually have some domain knowledge*.  The one key thing that I *always tried to put into practice*:

* Security should be an enabling technology.  It should not exist to tell people "no"-- it should work to make sure there is a reasonable, attainable path to meet the critical business or end-user need.  The moment that it stops doing so, is the moment that your users actively work against the security infrastructure you are trying to build and everything is lost.

**Comment by ocdtrekkie** [comment]
@mlyle Well...

> when you're talking to people that they may actually have some domain knowledge.

You may, but, to be honest, you really blew credibility out the window in your opening remarks. "Asking the user whether to do it is pretty damn safe" is a comment that really, absolutely is wholly ridiculous in itself, and I don't even know where to go from there. If this was the case, UAC would've entirely eliminated all malware on Windows platforms back when Vista came out.

> it should work to make sure there is a reasonable, attainable path to meet the critical business or end-user need

And indeed there is. There are two ways to do this I suggested without introducing a massive security hole into the browser ecosystem. Enabling users to accomplish their goals is absolutely a key aspect of providing a secure environment, but there's a difference between ensuring there's an attainable path, and just doing whatever the requestor wants you to do. There's an attainable path, it just doesn't involve the Serial API.

> I used to act like you sometimes

This is you trying to "put other people down for my point to stand". Maybe you haven't changed as much as you think.

> get back to debating the technical merits

For what it's worth, @SuperUserNameMan's last comment was excellent, and I will work on addressing it shortly! There's a *lot* to cover, so I need to put a fair bit together to respond adequately.

**Comment by mlyle** [comment]
@ocdtrekkie 

> You may, but, to be honest, you really blew credibility out the window in your opening remarks. 

You still refuse to refrain from personal attacks.  Address the issue, not the person.

**Comment by bgrins** [comment]
I've locked the thread for now. Many of these recent comments are well outside of the participation guidelines at https://github.com/mozilla/standards-positions/blob/main/CONTRIBUTING.md#for-the-broader-community. At this point the conversation should be focused on any changes to the spec since the position was set.
